### ISSUE TITLE
is_displayed

### ISSUE THREAD

**Issue by ylavoie** [teodesian/Selenium-Remote-Driver#386]
S::R::D documentation states:
```
 Description:
    Determine if an element is currently displayed.
    Note: This does *not* tell you an element's 'visibility' property; as it still takes up space in the DOM and is therefore considered 'displayed'.

 Output:
    BOOLEAN - Whether the element is displayed.

 Usage:
    $elem->is_displayed();
```
which is different from https://w3c.github.io/webdriver/#element-displayedness' which states:
```
The visibility of an element is guided by what is perceptually visible to the human eye. In this context, an element's displayedness does not relate to the visibility or display style properties.

The approach recommended to implementors to ascertain an element's visibility was originally developed by the Selenium project, and is based on crude approximations about an element's nature and relationship in the tree. An element is in general to be considered visible if any part of it is drawn on the canvas within the boundaries of the viewport.
```

**Comment by teodesian** [comment]
I tried to keep behaving like our existing methods which were based on JSONWire.  I believe under JSONWire protocol, this description is correct, but I'll have to look again at it.  It may indeed be that this method is now doing something different or vice versa, and things need to be adjusted.  I just need to look and see what particularly is the case here.

**Comment by ylavoie** [comment]
My understanding is that the actual implementation returns true if the element is in DOM but not shown.
Even if JSONWire doesn't clearly define what `displayed` means in https://github.com/SeleniumHQ/selenium/wiki/JsonWireProtocol#sessionsessionidelementiddisplayed, I think that it has to be shown in the current view port and not only present in DOM.

**Comment by Qarj** [comment]
This is the function (which I found on Stack Overflow) which I have been using to determine if the element is in the view port. It seems to work fairly well for us:

```
        function isElementInViewport (el) {
        
            var rect = el.getBoundingClientRect();
        
            return (
                rect.top >= 0 &&
                rect.left >= 0 &&
                rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) && /*or $(window).height() */
                rect.right <= (window.innerWidth || document.documentElement.clientWidth) /*or $(window).width() */
            );
        }
```

**Comment by ylavoie** [comment]
Qarj, thanks for the suggestion. The issue is not if the element is in the viewport but if `is_displayed` means present in the DOM or, like the W3C and JSONWire standards, if it means actually shown on the viewport.
I know that it could break some current implementation, but S::R::D is perl implementation of those standards and should be compliant with them.

**Comment by teodesian** [comment]
> I know that it could break some current implementation, but S::R::D is perl implementation of those standards and should be compliant with them.

I disagree.  Breaking people's tests is the last thing I want to do.  That said, I want there to be a way to accomplish what you want done.  Would it be acceptable to add a flag to control this behavior, or to pass an additional parameter to the call controlling whether we are sticking tightly to the spec?

**Comment by Qarj** [comment]
For the record it won't break many of our tests - I'm sure we could live with this. It would be a nice though if the current behaviour could be preserved - even in a different method name.

Extra care and effort is needed when writing assertions that something is in view port - especially on Windows since our view ports can change wildly - e.g. 1920 x 1080 or more when developing, then dropping to a max of 1024 x 768 when run using LocalSystem Account or a service account.

**Comment by teodesian** [comment]
Ok, I have finally had some time to look into the documentation on all this.

The JSONWire spec makes no attempt whatsoever to define what 'displayed' is.  In my experience this was merely whether the element had space allocated for it in the DOM, e.g. the whether style: display:none was applied.

The current polyfill for chrome and edge simply do this.  A case can certainly be made that in WD3 mode the polyfill should do what the WD3 method *is supposed to do*, as I'm making no attempt here to correct behaviour to the prior way when WD3 actually works.

I've heard no complaints from the peanut gallery out there about this state of affairs, so I'm now inclined to simply change the chrome polyfill to do the viewport check.

So, in summary I'm going to 

1. do the viewport check by default
2. allow an optional flag to be passed to do the "old style" only check that lamely only checks the attr.

**Comment by teodesian** [comment]
Actually, given the lack of complaint, A POD advisory should suffice.  I think I'm on to the solution.
